File: src/787_cheapest_flights_within_k_stops/cheapest_flights_within_k_stops.py
from collections import deque
from typing import List


class Solution:
    def findCheapestPrice(
        self, n: int, flights: List[List[int]], src: int, dst: int, k: int
    ) -> int:
        prices = [float("inf")] * n
        prices[src] = 0
        adjList = [[] for _ in range(n)]
        for u, v, cst in flights:
            adjList[u].append([v, cst])

        q = deque([(0, src, 0)])

        while q:
            cst, node, stops = q.popleft()
            if stops > k:
                continue

            for adj, adjCost in adjList[node]:
                newCost = cst + adjCost
                if newCost < prices[adj]:
                    prices[adj] = newCost
                    q.append((newCost, adj, stops + 1))

        return prices[dst] if prices[dst] != float("inf") else -1


# Test helper
def test_find_cheapest_price():
    solution = Solution()

    # Test case 1: Example 1
    n1 = 4
    flights1 = [[0, 1, 100], [1, 2, 100], [2, 0, 100], [1, 3, 600], [2, 3, 200]]
    src1, dst1, k1 = 0, 3, 1
    expected1 = 700
    actual1 = solution.findCheapestPrice(n1, flights1, src1, dst1, k1)
    print(f"{actual1} | {expected1}")

    # Test case 2: Example 2
    n2 = 3
    flights2 = [[0, 1, 100], [1, 2, 100], [0, 2, 500]]
    src2, dst2, k2 = 0, 2, 1
    expected2 = 200
    actual2 = solution.findCheapestPrice(n2, flights2, src2, dst2, k2)
    print(f"{actual2} | {expected2}")

    # Test case 3: Example 3
    n3 = 3
    flights3 = [[0, 1, 100], [1, 2, 100], [0, 2, 500]]
    src3, dst3, k3 = 0, 2, 0
    expected3 = 500
    actual3 = solution.findCheapestPrice(n3, flights3, src3, dst3, k3)
    print(f"{actual3} | {expected3}")

    # Test case 4: No path
    n4 = 3
    flights4 = [[0, 1, 100], [1, 2, 100]]
    src4, dst4, k4 = 0, 2, 0
    expected4 = -1
    actual4 = solution.findCheapestPrice(n4, flights4, src4, dst4, k4)
    print(f"{actual4} | {expected4}")

    # Test case 5: Direct flight cheaper
    n5 = 2
    flights5 = [[0, 1, 50], [0, 1, 100]]
    src5, dst5, k5 = 0, 1, 1
    expected5 = 50
    actual5 = solution.findCheapestPrice(n5, flights5, src5, dst5, k5)
    print(f"{actual5} | {expected5}")


if __name__ == "__main__":
    test_find_cheapest_price()

File: src/787_cheapest_flights_within_k_stops/cheapest_flights_within_k_stops.cpp
#include <vector>

class Solution {
   public:
    int findCheapestPrice(int n, vector<vector<int>>& flights, int src, int dst, int k) {
        vector<int> prices(n, INT_MAX);
        prices[src] = 0;

        vector<vector<pair<int, int>>> adjList(n);
        for (auto& flight : flights) {
            int u = flight[0], v = flight[1], cst = flight[2];
            adjList[u].push_back({v, cst});
        }

        queue<tuple<int, int, int>> q;  // {cost, node, stops}
        q.push({0, src, 0});

        while (!q.empty()) {
            auto [cst, node, stops] = q.front();
            q.pop();
            if (stops > k) continue;

            for (auto& adj : adjList[node]) {
                int adjNode = adj.first, adjCost = adj.second;
                int newCost = cst + adjCost;
                if (newCost < prices[adjNode]) {
                    prices[adjNode] = newCost;
                    q.push({newCost, adjNode, stops + 1});
                }
            }
        }

        return prices[dst] == INT_MAX ? -1 : prices[dst];
    }
};

// Test helper
void testFindCheapestPrice() {
    Solution solution;

    // Test case 1: Example 1
    int n1 = 4;
    std::vector<std::vector<int>> flights1 = {
        {0, 1, 100}, {1, 2, 100}, {2, 0, 100}, {1, 3, 600}, {2, 3, 200}};
    int src1 = 0, dst1 = 3, k1 = 1;
    int expected1 = 700;
    int actual1 = solution.findCheapestPrice(n1, flights1, src1, dst1, k1);
    std::cout << actual1 << " | " << expected1 << std::endl;

    // Test case 2: Example 2
    int n2 = 3;
    std::vector<std::vector<int>> flights2 = {{0, 1, 100}, {1, 2, 100}, {0, 2, 500}};
    int src2 = 0, dst2 = 2, k2 = 1;
    int expected2 = 200;
    int actual2 = solution.findCheapestPrice(n2, flights2, src2, dst2, k2);
    std::cout << actual2 << " | " << expected2 << std::endl;

    // Test case 3: Example 3
    int n3 = 3;
    std::vector<std::vector<int>> flights3 = {{0, 1, 100}, {1, 2, 100}, {0, 2, 500}};
    int src3 = 0, dst3 = 2, k3 = 0;
    int expected3 = 500;
    int actual3 = solution.findCheapestPrice(n3, flights3, src3, dst3, k3);
    std::cout << actual3 << " | " << expected3 << std::endl;

    // Test case 4: No path
    int n4 = 3;
    std::vector<std::vector<int>> flights4 = {{0, 1, 100}, {1, 2, 100}};
    int src4 = 0, dst4 = 2, k4 = 0;
    int expected4 = -1;
    int actual4 = solution.findCheapestPrice(n4, flights4, src4, dst4, k4);
    std::cout << actual4 << " | " << expected4 << std::endl;

    // Test case 5: Direct flight cheaper
    int n5 = 2;
    std::vector<std::vector<int>> flights5 = {{0, 1, 50}, {0, 1, 100}};
    int src5 = 0, dst5 = 1, k5 = 1;
    int expected5 = 50;
    int actual5 = solution.findCheapestPrice(n5, flights5, src5, dst5, k5);
    std::cout << actual5 << " | " << expected5 << std::endl;
}

int main() {
    testFindCheapestPrice();
    return 0;
}
